File: projects/swagger-studio-ruleset/python/src/swagger_studio_ruleset_publisher/activator.py
"""Mark an uploaded Spectral ruleset as the org's active style guide
— or revoke that mark.

`spectral:upload` only saves a draft style guide; the SwaggerHub UI's
"Publish" button is what flips `spectralRulesets[].enabled` in the org's
standardization config. Without an enabled entry, `/standardization`
returns no findings — the engine has nothing to evaluate against.

Flow (works around the field-naming asymmetry SwaggerHub uses — upload
addresses rulesets by name, the org config addresses them by UUID):

  1. (skippable) GET /standardization/spectral-rulesets/{owner}
       -> array of {id, name, ...}. Look up the UUID by `name`.
       Retried with backoff to ride out the indexing lag right after upload.
       SKIPPED when the caller already has the UUID (the REST upload
       endpoint returns it in the response body — pass it via the
       `ruleset_id` arg to short-circuit this step).

  2. GET  /standardization/{owner}/config
       -> shallow form: spectralRulesets[] entries carry only
       `rulesetId` (the UUID from step 1) and `enabled`.

  3. POST /standardization/{owner}/config  (entire modified body)
       Find the entry whose `rulesetId` == lookup UUID, flip `enabled`.
       For `activate`: create the entry if missing. For `deactivate`:
       if there's no entry there's nothing to disable — return early.

Endpoints discovered by capturing the UI's network calls.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any

import httpx
import structlog

from swagger_studio_ruleset_publisher._http import create_client
from swagger_studio_ruleset_publisher.config import Settings

log = structlog.get_logger(__name__)

# Backoff schedule for the lookup step. Total wait <= 7s.
_LOOKUP_DELAYS_S: tuple[float, ...] = (0.0, 2.0, 5.0)


class RulesetNotFoundError(RuntimeError):
    """The named ruleset wasn't found in the org's rulesets list."""


@dataclass(frozen=True)
class ActivationResult:
    owner: str
    ruleset_name: str
    ruleset_id: str
    enabled: bool
    studio_url: str
    detail: str


async def activate(
    settings: Settings,
    ruleset_name: str,
    ruleset_id: str | None = None,
) -> ActivationResult:
    """Flip enabled=True for `ruleset_name` in the org config and persist.

    If `ruleset_id` is supplied (the REST upload returns it), the lookup
    round-trip is skipped — saves one request and removes the
    indexing-lag retry surface. Falls back to lookup-by-name otherwise.
    """
    return await _set_state(settings, ruleset_name, enabled=True, ruleset_id=ruleset_id)


async def deactivate(
    settings: Settings,
    ruleset_name: str,
    ruleset_id: str | None = None,
) -> ActivationResult:
    """Flip enabled=False for `ruleset_name` in the org config and persist.

    Idempotent: if no entry exists in the config, returns success with
    `detail` noting the no-op. The ruleset itself isn't deleted —
    its content stays in Studio under the original slot.
    """
    return await _set_state(settings, ruleset_name, enabled=False, ruleset_id=ruleset_id)


async def _set_state(
    settings: Settings,
    ruleset_name: str,
    *,
    enabled: bool,
    ruleset_id: str | None,
) -> ActivationResult:
    """Shared flow for activate/deactivate. `enabled` decides the direction."""
    owner = settings.swaggerhub_org
    op = "activate" if enabled else "deactivate"

    async with create_client(settings) as client:
        if ruleset_id is None:
            ruleset_id = await lookup_ruleset_id(client, owner, ruleset_name)
            if ruleset_id is None:
                raise RulesetNotFoundError(
                    f"Ruleset {ruleset_name!r} not found in /standardization/spectral-rulesets/{owner} "
                    f"after {len(_LOOKUP_DELAYS_S)} attempts."
                )
        else:
            log.info(f"{op}_lookup_skipped", ruleset_id=ruleset_id)

        config_path = f"/standardization/{owner}/config"
        log.info(f"{op}_get", path=config_path, ruleset_id=ruleset_id)
        get_resp = await client.get(config_path)
        get_resp.raise_for_status()
        config: dict[str, Any] = get_resp.json()

        changed = _set_enabled_by_id(config, ruleset_id, enabled=enabled)
        if not changed:
            # Deactivating something that was never in the config — no-op.
            return ActivationResult(
                owner=owner,
                ruleset_name=ruleset_name,
                ruleset_id=ruleset_id,
                enabled=False,
                studio_url=f"https://app.swaggerhub.com/organizations/{owner}/governance",
                detail="no-op (entry not present in config)",
            )

        log.info(f"{op}_post", path=config_path, ruleset_id=ruleset_id)
        post_resp = await client.post(
            config_path, json=config, headers={"Content-Type": "application/json"}
        )
        if post_resp.status_code >= 400:
            raise RuntimeError(
                f"{op.capitalize()} POST returned HTTP {post_resp.status_code}: "
                f"{post_resp.text.strip()}"
            )

    return ActivationResult(
        owner=owner,
        ruleset_name=ruleset_name,
        ruleset_id=ruleset_id,
        enabled=enabled,
        studio_url=f"https://app.swaggerhub.com/organizations/{owner}/governance",
        detail=f"HTTP {post_resp.status_code}",
    )


async def lookup_ruleset_id(
    client: httpx.AsyncClient,
    owner: str,
    ruleset_name: str,
) -> str | None:
    """Find the ruleset's UUID by name. Retries to ride out upload-indexing lag.

    Public so the deleter / other modules can reuse it.
    """
    path = f"/standardization/spectral-rulesets/{owner}"
    last_status: int | None = None
    for delay in _LOOKUP_DELAYS_S:
        if delay > 0:
            log.info("lookup_retry", delay_s=delay, path=path)
            await asyncio.sleep(delay)
        else:
            log.info("lookup", path=path)
        resp = await client.get(path)
        last_status = resp.status_code
        if resp.status_code >= 400:
            continue
        data = resp.json()
        if not isinstance(data, list):
            continue
        for entry in data:
            if not isinstance(entry, dict):
                continue
            if entry.get("name") == ruleset_name:
                rid = entry.get("id")
                if isinstance(rid, str) and rid:
                    return rid
    log.warning("lookup_exhausted", last_status=last_status)
    return None


def _set_enabled_by_id(
    config: dict[str, Any], ruleset_id: str, enabled: bool = True
) -> bool:
    """Set the entry's `enabled` flag. Returns True if a write is needed.

    For `enabled=True`: appends a new entry if one doesn't exist (activate
    semantics — caller wants this thing turned on regardless of prior state).

    For `enabled=False`: returns False when no entry exists (deactivate
    semantics — nothing to disable, and creating a disabled entry would
    be pointless noise).

    Mutates `config` in place when applicable.
    """
    rulesets = config.get("spectralRulesets")
    if not isinstance(rulesets, list):
        if not enabled:
            return False
        rulesets = []
        config["spectralRulesets"] = rulesets

    for entry in rulesets:
        if isinstance(entry, dict) and entry.get("rulesetId") == ruleset_id:
            entry["enabled"] = enabled
            return True

    if not enabled:
        return False

    rulesets.append({"rulesetId": ruleset_id, "enabled": True})
    return True


def remove_entry_by_id(config: dict[str, Any], ruleset_id: str) -> bool:
    """Drop the `spectralRulesets[]` entry for `ruleset_id`. Returns True if removed.

    Used by the deleter after a successful DELETE so we don't leave a
    dangling rulesetId in the config pointing at a slot that no longer
    exists.
    """
    rulesets = config.get("spectralRulesets")
    if not isinstance(rulesets, list):
        return False
    new_list = [
        e
        for e in rulesets
        if not (isinstance(e, dict) and e.get("rulesetId") == ruleset_id)
    ]
    if len(new_list) == len(rulesets):
        return False
    config["spectralRulesets"] = new_list
    return True

File: projects/swagger-studio-ruleset/typescript/src/activator.ts
/**
 * Mark an uploaded Spectral ruleset as the org's active style guide —
 * or revoke that mark.
 *
 * Flow (works around SwaggerHub's field-naming asymmetry — upload addresses
 * rulesets by name, the org config addresses them by UUID):
 *
 *   1. (skippable) GET /standardization/spectral-rulesets/{owner}
 *        -> array of {id, name, ...}. Look up the UUID by `name`.
 *        Retried with backoff to ride out the post-upload indexing lag.
 *        SKIPPED when the caller already has the UUID.
 *
 *   2. GET  /standardization/{owner}/config
 *        -> spectralRulesets[] entries carry {rulesetId, enabled}.
 *
 *   3. POST /standardization/{owner}/config  (entire modified body)
 *        For activate: flip enabled=true, append if missing.
 *        For deactivate: flip enabled=false; if no entry, no-op.
 */

import type { Settings } from "./config.js";
import { buildHeaders, studioBaseUrl, withTimeout } from "./_http.js";

export interface ActivationResult {
  readonly owner: string;
  readonly rulesetName: string;
  readonly rulesetId: string;
  readonly enabled: boolean;
  readonly studioUrl: string;
  readonly detail: string;
}

export class RulesetNotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "RulesetNotFoundError";
  }
}

const LOOKUP_DELAYS_MS = [0, 2_000, 5_000] as const;

export async function activate(
  settings: Settings,
  rulesetName: string,
  knownRulesetId?: string,
): Promise<ActivationResult> {
  return setState(settings, rulesetName, true, knownRulesetId);
}

export async function deactivate(
  settings: Settings,
  rulesetName: string,
  knownRulesetId?: string,
): Promise<ActivationResult> {
  return setState(settings, rulesetName, false, knownRulesetId);
}

async function setState(
  settings: Settings,
  rulesetName: string,
  enabled: boolean,
  knownRulesetId?: string,
): Promise<ActivationResult> {
  const owner = settings.swaggerhubOrg;
  const baseUrl = studioBaseUrl(settings);
  const headers = buildHeaders(settings, { "Content-Type": "application/json" });

  return withTimeout(settings, async (signal) => {
    let rulesetId: string | null;
    if (knownRulesetId !== undefined && knownRulesetId.length > 0) {
      rulesetId = knownRulesetId;
    } else {
      rulesetId = await lookupRulesetId(baseUrl, owner, rulesetName, headers, signal);
      if (rulesetId === null) {
        throw new RulesetNotFoundError(
          `Ruleset '${rulesetName}' not found in /standardization/spectral-rulesets/${owner} ` +
            `after ${String(LOOKUP_DELAYS_MS.length)} attempts.`,
        );
      }
    }

    const configUrl = `${baseUrl}/standardization/${owner}/config`;

    const getResp = await fetch(configUrl, { method: "GET", headers, signal });
    if (!getResp.ok) {
      const body = await getResp.text();
      throw new Error(`GET config returned HTTP ${String(getResp.status)}: ${body.trim()}`);
    }
    const config = (await getResp.json()) as Record<string, unknown>;

    const changed = setEnabledById(config, rulesetId, enabled);
    if (!changed) {
      // Deactivating something not in the config — no-op.
      return {
        owner,
        rulesetName,
        rulesetId,
        enabled: false,
        studioUrl: `https://app.swaggerhub.com/organizations/${owner}/governance`,
        detail: "no-op (entry not present in config)",
      };
    }

    const postResp = await fetch(configUrl, {
      method: "POST",
      headers,
      body: JSON.stringify(config),
      signal,
    });
    if (!postResp.ok) {
      const body = await postResp.text();
      throw new Error(`POST config returned HTTP ${String(postResp.status)}: ${body.trim()}`);
    }

    return {
      owner,
      rulesetName,
      rulesetId,
      enabled,
      studioUrl: `https://app.swaggerhub.com/organizations/${owner}/governance`,
      detail: `HTTP ${String(postResp.status)}`,
    };
  });
}

/**
 * Look up the ruleset's UUID by name. Retries to ride out upload-indexing lag.
 * Exported so deleter and other modules can reuse.
 */
export async function lookupRulesetId(
  baseUrl: string,
  owner: string,
  rulesetName: string,
  headers: Record<string, string>,
  signal: AbortSignal,
): Promise<string | null> {
  const url = `${baseUrl}/standardization/spectral-rulesets/${owner}`;
  for (const delay of LOOKUP_DELAYS_MS) {
    if (delay > 0) {
      await new Promise<void>((res) => setTimeout(res, delay));
    }
    const resp = await fetch(url, { method: "GET", headers, signal });
    if (!resp.ok) continue;
    const data: unknown = await resp.json();
    if (!Array.isArray(data)) continue;
    for (const entry of data) {
      if (isRecord(entry) && entry["name"] === rulesetName) {
        const id = entry["id"];
        if (typeof id === "string" && id.length > 0) {
          return id;
        }
      }
    }
  }
  return null;
}

/**
 * Set the entry's `enabled` flag. Returns true if the config was modified.
 *
 * Backward-compatible: defaults to `enabled=true` so existing call sites
 * (and tests) keep working. For deactivate semantics (`enabled=false`),
 * a missing entry yields `false` — there's nothing to disable, and creating
 * a disabled entry would be pointless noise in the config.
 */
export function setEnabledById(
  config: Record<string, unknown>,
  rulesetId: string,
  enabled = true,
): boolean {
  let rulesets = config["spectralRulesets"];
  if (!Array.isArray(rulesets)) {
    if (!enabled) return false;
    rulesets = [];
    config["spectralRulesets"] = rulesets;
  }
  const arr = rulesets as unknown[];
  for (const entry of arr) {
    if (isRecord(entry) && entry["rulesetId"] === rulesetId) {
      entry["enabled"] = enabled;
      return true;
    }
  }
  if (!enabled) return false;
  arr.push({ rulesetId, enabled: true });
  return true;
}

/**
 * Drop the spectralRulesets[] entry for `rulesetId`. Returns true if removed.
 * Used by the deleter after a successful DELETE so we don't leave a
 * dangling rulesetId in the config.
 */
export function removeEntryById(config: Record<string, unknown>, rulesetId: string): boolean {
  const rulesets = config["spectralRulesets"];
  if (!Array.isArray(rulesets)) return false;
  const arr = rulesets as unknown[];
  const next = arr.filter(
    (e) => !(isRecord(e) && e["rulesetId"] === rulesetId),
  );
  if (next.length === arr.length) return false;
  config["spectralRulesets"] = next;
  return true;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
